rust-compile-time-adventures-with-tikv-ep1.md

```diff
@@ -0,0 +1,267 @@
+>
+>* _7 femto-bunnies_ - rustboot building Rust prior to being retired
+>* _49 kilo-hamsters_ - rustc building Rust immediately after rustboot’s retirement
+>* _188 giga-sloths_ - rustc building Rust in 2020
+
+Anyway, last time I bootstrapped Rust a few months ago, it took over five hours.
+
+The Rust language developers became acclimated to Rust's poor self-hosting times and failed to recognize or address the severity of the problem of bad compile times during Rust's crucial early design phase.
+
+### (Un)virtuous cycles
+
+In the Rust project, we like processes that reinforce and build upon themselves. This is one of the keys to Rust's success, both as a language and a community.
+
+As an obvious, hugely-successful example, consider [Servo](https://github.com/servo/servo). Servo is a web browser built in Rust, and Rust was created with the explicit purpose of building Servo. Rust and Servo are sister-projects. They were created by the same team (initially), at roughly the same time, and they evolved together. Not only was Rust built to create Servo, but Servo was built to inform the design of Rust.
+
+The initial few years of both projects were extremely difficult, with both projects evolving in parallel. The often-used metaphor of the [Ship of Theseus](https://en.wikipedia.org/wiki/Ship_of_Theseus) is apt - we were constantly rebuilding Rust in order to sail the seas of Servo. There is no doubt that the experience of building Servo with Rust while simultaneously building the language itself led directly to many of the good decisions that make Rust the practical language it is.
+
+Here are some cursory examples of the Servo-Rust feedback loop:
+
+* Labeled break and continue [was implemented in order to auto-generate an HTML parser](https://github.com/rust-lang/rust/issues/2216).
+* Owned closures [were implemented after analyzing closure usage in Servo](https://github.com/rust-lang/rust/issues/2549#issuecomment-19588158).
+* External function calls used to be considered safe. [This changed in part due to experience in Servo](https://github.com/rust-lang/rust/issues/2628#issuecomment-9384243).
+* The migration from green-threading to native threading was informed by the experience of building Servo, observing the FFI overhead of Servo's SpiderMonkey integration, and profiling "hot splits", where the green thread stacks needed to be expanded and contracted.
+
+The co-development of Rust and Servo created a [virtuous cycle](https://en.wikipedia.org/wiki/Virtuous_circle_and_vicious_circle) that allowed both projects to thrive. Today, Servo components are deeply integrated into Firefox, ensuring that Rust cannot die while Firefox lives.
+
+Mission accomplished.
+
+![Rust mission accomplished](media/rust-compile-mission-completed.png)
```

Comment:
Is this image from the Internet source? If yes, we need to add the source here to avoid copyright issues.

Reply:
The image comes from NASA so I believe it is in the public domain.

I downloaded it here: https://www.bbc.co.uk/radioscotland/60s/moonlandings/05

Reply:
Nice photoshop there!